net/tcp/tcp_recvfrom.c

```diff
@@ -843,10 +843,74 @@ ssize_t psock_tcp_recvfrom(FAR struct socket *psock, FAR struct msghdr *msg,
     }
 
   tcp_notify_recvcpu(conn);
+  tcp_recvfrom_uninitialize(&state);
+  return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: psock_tcp_recvfrom
+ *
+ * Description:
+ *   Perform the recvfrom operation for a TCP/IP SOCK_STREAM
+ *
+ * Input Parameters:
+ *   psock    Pointer to the socket structure for the SOCK_DRAM socket
+ *   msg      Receive info and buffer for receive data
+ *   flags    Receive flags
+ *
+ * Returned Value:
+ *   On success, returns the number of characters received.  On  error,
+ *   -errno is returned (see recvfrom for list of errnos).
+ *
+ * Assumptions:
+ *
+ ****************************************************************************/
+
+ssize_t psock_tcp_recvfrom(FAR struct socket *psock, FAR struct msghdr *msg,
+                           int flags)
+{
+  FAR struct sockaddr   *from    = msg->msg_name;
+  FAR socklen_t         *fromlen = &msg->msg_namelen;
+  FAR struct tcp_conn_s *conn;
+  ssize_t                nrecv   = 0;
+  ssize_t                ret     = 0;
+  int                    i;
+
+  net_lock();
+
+  conn = psock->s_conn;
+  for (i = 0; i < msg->msg_iovlen; i++)
+    {
+      FAR void *buf = msg->msg_iov[i].iov_base;
+      size_t len = msg->msg_iov[i].iov_len;
+
+      ret = tcp_recvfrom_one(conn, buf, len, from, fromlen, flags);
+      if (ret <= 0)
+        {
+          break;
+        }
+
+      nrecv += ret;
+
+      /* User has not set MSG_WAITALL: If the first buffer is full
+       * when received for the first time, then check the next buffer,
+       * otherwise return the received data.
+       * User sets MSG_WAITALL: Ensure that each iov is filled before
+       * returning
+       */
+
+      if (!(flags & MSG_WAITALL) && ret < msg->msg_iov[i].iov_len)
```

Comment:
If MSG_WAITALL is true, the offset of the currently received data should also be updated here instead of skipping the current slot reading

Reply:
could you share where do you find the required action?

Reply:
I'm only discussing the current code snippet, which has this risk